cmd/ipfs/daemon.go

```diff
@@ -244,48 +222,73 @@ func daemonFunc(req cmds.Request, res cmds.Response) {
 	}
 
 	// construct api endpoint - every time
-	err, apiErrc := serveHTTPApi(req)
+	err, apiErrc := serveHTTPApi(req, node)
 	if err != nil {
 		res.SetError(err, cmds.ErrNormal)
 		return
 	}
 
 	// construct http gateway - if it is set in the config
-	var gwErrc <-chan error
-	if len(cfg.Addresses.Gateway) > 0 {
-		var err error
-		err, gwErrc = serveHTTPGateway(req)
-		if err != nil {
-			res.SetError(err, cmds.ErrNormal)
-			return
-		}
+	err, gwErrc := maybeServeHTTPGateway(req, node)
+	if err != nil {
+		res.SetError(err, cmds.ErrNormal)
+		return
 	}
 
 	// construct fuse mountpoints - if the user provided the --mount flag
-	mount, _, err := req.Option(mountKwd).Bool()
+	err, mountErrc := maybeMountFuse(req, node)
 	if err != nil {
 		res.SetError(err, cmds.ErrNormal)
 		return
 	}
-	if mount {
-		if err := mountFuse(req); err != nil {
-			res.SetError(err, cmds.ErrNormal)
-			return
-		}
+
+	// repo blockstore GC - if --disable-gc flag is not present
+	err, gcErrc := maybeRunGC(req, node)
+	if err != nil {
+		res.SetError(err, cmds.ErrNormal)
+		return
 	}
 
 	// collect long-running errors and block for shutdown
-	// TODO(cryptix): our fuse currently doesnt follow this pattern for graceful shutdown
-	for err := range merge(apiErrc, gwErrc) {
+	var errs []error
+	for err := range merge(apiErrc, gwErrc, mountErrc, gcErrc) {
 		if err != nil {
-			res.SetError(err, cmds.ErrNormal)
-			return
+			log.Error(err)
+			errs = append(errs, err)
 		}
 	}
+	if len(errs) > 0 {
+		errStr := ""
+		for _, e := range errs {
+			errStr += " " + e.Error()
```

Comment:
I normally just return the first error i get and dont bother trying to combine them. Although definitely log each one